### PR TITLE
Added transit/json formatting and (de)serialization helpers.

### PR DESCRIPTION
**Pull request by ghost**
Messages can now be sent and received with JSON and Transit (default) encoded bodies. There's also an option left to send raw messages without converting the body payload.

Updated the existing tests accordingly and added new tests for the (de)serialization helpers.

### CODE REVIEW

File: src/clj_sqs_extended/serdes.clj [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+  (:require [cognitect.transit :as transit]
+            [tick.alpha.api :as t]
+            [cheshire.core :as json])
+  (:import [java.io ByteArrayOutputStream ByteArrayInputStream]))
+
+
+(def ^:private tick-time-writer
+  (transit/write-handler
+    "tick-time"
+    (fn [t] (.toString t))))
```
Comment (Quantisan) [inline src/clj_sqs_extended/serdes.clj]: looks like there's an outstanding PR upstream to enable serde for tick <> transit.  https://github.com/juxt/tick/issues/34#issuecomment-503641158, which points to https://github.com/henryw374/time-literals/issues/2

perhaps make use of their suggestion https://gist.github.com/jdf-id-au/2e91fb63ce396b722c1d6770154f1815#file-transit-connection-cljc-L54 as it handles all the tick classes

File: src/clj_sqs_extended/serdes.clj [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+                      {:handlers {"tick-time" tick-time-reader}})
+      (transit/read)))
+
+(defn serialize
+  [out format]
+  (cond
+    (= format :transit) (transit-write out)
+    (= format :json) (json/generate-string out)
+    (= format :raw) out
+    :else out))
```
Comment (Quantisan) [inline src/clj_sqs_extended/serdes.clj]: if format is not known, perhaps throw an appropriate exception? e.g. `ex-info`

File: test/clj_sqs_extended/serdes_test.clj [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+(def ^:private basic-map
+  {:quote "State. You're doing it wrong."
+   :by "Rich Hickey"})
+
+(def ^:private timestamp-map
+  {:step 1
+   :timestamp (t/inst)})
+
+(deftest roundtrip-transit
+  (testing "Transit with basic map"
```
Comment (Quantisan) [inline test/clj_sqs_extended/serdes_test.clj]: I'd refactor up these different tests into their own top level tests. i.e. `roundtrip-transit-with-basic-map` and `roundtrip-transit-with-timestamp`

File: test/clj_sqs_extended/serdes_test.clj [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+             (serdes/serialize basic-map :transit)
+             :transit))))
+  (testing "Transit with timestamp"
+    (is (= timestamp-map
+           (serdes/deserialize
+             (serdes/serialize timestamp-map :transit)
+             :transit)))))
+
+(deftest roundtrip-json
+  (testing "JSON with basic map"
```
Comment (Quantisan) [inline test/clj_sqs_extended/serdes_test.clj]: what about round trip with timestamp like above?

File: test/clj_sqs_extended/core_test.clj [hunk 5]
```diff
@@ -37,19 +45,22 @@
...
-        (is (= test-message-2 (:body message)))))))
+      (doseq [format [:transit :json]]
+        (let [group-id-1 (str format "-1")
+              group-id-2 (str format "-2")]
+          (sqs-ext/send-fifo-message client url test-message-1 group-id-1 {:format format})
+          (sqs-ext/send-fifo-message client url test-message-2 group-id-2 {:format format}))
+        (let [message (sqs-ext/receive-message client url {:format format})]
+          (is (= test-message-1 (:body message))))
+        (let [message (sqs-ext/receive-message client url {:format format})]
+          (is (= test-message-2 (:body message))))))))
```
Comment (Quantisan) [inline test/clj_sqs_extended/core_test.clj]: are 2 messages sufficient entropy to guarantee that the ordering works?

File: src/clj_sqs_extended/core.clj [hunk 6]
```diff
@@ -72,43 +73,58 @@
...
     (.deleteMessage sqs-client request)))
 
 (defn receive-message
   ([sqs-client url]
    (receive-message sqs-client url {}))
   ([sqs-client url
-    {:keys [wait-time
+    {:keys [format
+            wait-time
             visibility-timeout
```
Comment (Quantisan) [inline src/clj_sqs_extended/core.clj]: can we remove the visibility-timeout setting here [as noted last time](https://github.com/Motiva-AI/clj-sqs-extended/pull/9#discussion_r461358206)? this is a configuration feature that I want to remove from this library. Or, are you leaving this here deliberately?

File: test/clj_sqs_extended/core_test.clj [hunk 7]
```diff
@@ -58,10 +69,12 @@
...
     (let [client (fixtures/localstack-sqs)
           url (fixtures/test-queue-url)]
-      (sqs-ext/send-message client url test-message-1)
-      (let [message (sqs-ext/receive-message client url {:visibility-timeout 1})]
-        (is (= test-message-1 (:body message))))
-      (Thread/sleep 2000)
-      (let [message (sqs-ext/receive-message client url)]
-        (is (= test-message-1 (:body message)))))))
-
+      (doseq [format [:transit :json]]
```
Comment (Quantisan) [inline test/clj_sqs_extended/core_test.clj]: it's nice to have this for coverage. i'm on the fence about this though as we're testing functionality of SQS which perhaps we can assume they'd work as advertised. for sake of minimizing codebase.

File: src/clj_sqs_extended/core.clj [hunk 8]
```diff
@@ -72,43 +73,58 @@
...
-    (.sendMessage sqs-client request)))
+  ([sqs-client url body group-id]
+   (send-fifo-message sqs-client url body group-id {}))
+  ([sqs-client url body group-id
+    {:keys [format]
+     :or   {format :transit}
+     :as   opts}]
+   (let [payload (serdes/serialize body format)
+         request (SendMessageRequest. url payload)]
+     ; WATCHOUT: The group ID is mandatory when sending fifo messages.
```
Comment (Quantisan) [inline src/clj_sqs_extended/core.clj]: ❤️ on noting this not-obvious trap

File: src/clj_sqs_extended/core.clj [hunk 9]
```diff
@@ -72,43 +73,58 @@
...
-    ; WATCHOUT: The group ID is mandatory when sending fifo messages.
-    (doto request (.setMessageGroupId group-id))
-    (.sendMessage sqs-client request)))
+  ([sqs-client url body group-id]
+   (send-fifo-message sqs-client url body group-id {}))
+  ([sqs-client url body group-id
+    {:keys [format]
+     :or   {format :transit}
+     :as   opts}]
+   (let [payload (serdes/serialize body format)
```
Comment (Quantisan) [inline src/clj_sqs_extended/core.clj]: minor style note, since `payload` isn't used, why not `->>` this through to `SendMessageRequest.`?

File: src/clj_sqs_extended/core.clj [hunk 10]
```diff
@@ -72,43 +73,58 @@
...
-  (let [request (SendMessageRequest. url message)]
-    ; WATCHOUT: The group ID is mandatory when sending fifo messages.
-    (doto request (.setMessageGroupId group-id))
-    (.sendMessage sqs-client request)))
+  ([sqs-client url body group-id]
+   (send-fifo-message sqs-client url body group-id {}))
+  ([sqs-client url body group-id
+    {:keys [format]
+     :or   {format :transit}
+     :as   opts}]
```
Comment (Quantisan) [inline src/clj_sqs_extended/core.clj]: opts isn't used. you can leave it out.

File: src/clj_sqs_extended/core.clj [hunk 11]
```diff
@@ -72,43 +73,58 @@
...
 (defn send-message
-  [sqs-client url message]
-  (.sendMessage sqs-client url message))
+  ([sqs-client url body]
+   (send-message sqs-client url body {}))
+  ([sqs-client url body
+    {:keys [format]
+     :or   {format :transit}
+     :as   opts}]
+   (let [payload (serdes/serialize body format)]
```
Comment (Quantisan) [inline src/clj_sqs_extended/core.clj]: same here, `payload` is only used once, so just thread `->>` this through and save a `let`. I get that sometimes it's useful for readability. In this case though, `body` is obvious enough already.

File: src/clj_sqs_extended/core.clj [hunk 12]
```diff
@@ -72,43 +73,58 @@
   (.deleteQueue sqs-client url))
 
 (defn send-message
-  [sqs-client url message]
-  (.sendMessage sqs-client url message))
+  ([sqs-client url body]
+   (send-message sqs-client url body {}))
+  ([sqs-client url body
```
Comment (Quantisan) [inline src/clj_sqs_extended/core.clj]: I like to leave a blank line above here between the diff signatures

File: src/clj_sqs_extended/core.clj [hunk 13]
```diff
@@ -72,43 +73,58 @@
...
+         request (SendMessageRequest. url payload)]
+     ; WATCHOUT: The group ID is mandatory when sending fifo messages.
+     (doto request (.setMessageGroupId group-id))
+     (.sendMessage sqs-client request))))
 
 (defn delete-message
-  [sqs-client url message]
-  (let [request (DeleteMessageRequest. url (:receiptHandle message))]
+  [sqs-client url handle]
+  (let [request (DeleteMessageRequest. url (:receiptHandle handle))]
```
Comment (Quantisan) [inline src/clj_sqs_extended/core.clj]: again, just thread `request` through.

File: src/clj_sqs_extended/core.clj [hunk 14]
```diff
@@ -72,43 +73,58 @@
   (.deleteQueue sqs-client url))
 
 (defn send-message
-  [sqs-client url message]
-  (.sendMessage sqs-client url message))
+  ([sqs-client url body]
```
Comment (Quantisan) [inline src/clj_sqs_extended/core.clj]: I can see why you renamed the `message` arg to `body`. since there's an overloading of the term `message` between here vs the java client. however, it would read better if for a fn `send-message` to have an argument named `message` so that it's obvious to the user that we're sending this var, which contains the data that's to be sent. so I suggest sticking with naming this argument `message`.

File: src/clj_sqs_extended/core.clj [hunk 15]
```diff
@@ -72,43 +73,58 @@
...
+   (let [payload (serdes/serialize body format)]
+     (.sendMessage sqs-client url payload))))
 
 (defn send-fifo-message
-  [sqs-client url message group-id]
-  (let [request (SendMessageRequest. url message)]
-    ; WATCHOUT: The group ID is mandatory when sending fifo messages.
-    (doto request (.setMessageGroupId group-id))
-    (.sendMessage sqs-client request)))
+  ([sqs-client url body group-id]
```
Comment (Quantisan) [inline src/clj_sqs_extended/core.clj]: same here as `send-message` to rename `body` back to `message`

File: src/clj_sqs_extended/core.clj [hunk 16]
```diff
@@ -72,43 +73,58 @@
...
+   (let [payload (serdes/serialize body format)
+         request (SendMessageRequest. url payload)]
+     ; WATCHOUT: The group ID is mandatory when sending fifo messages.
+     (doto request (.setMessageGroupId group-id))
+     (.sendMessage sqs-client request))))
 
 (defn delete-message
-  [sqs-client url message]
-  (let [request (DeleteMessageRequest. url (:receiptHandle message))]
+  [sqs-client url handle]
```
Comment (Quantisan) [inline src/clj_sqs_extended/core.clj]: why is it called `handle`? it's not clear to the user what is a handle. I can see that you're extracting the `handle` from a mesage. But what's get passed is a message. And calling it handle is leaking implementation to the user.

lastly, there's an obvious pattern to the API signature. `send-mesage(client url message)`, so we should stick with `delete-message(client url message)` for consistency of the API terminology for user's sake.

**Review by Quantisan — CHANGES_REQUESTED** [review]
Clear and concise PR description that gives reviewer useful context. ✅ 
Straightforward implementation ✅ 
Kept to the intended scope of the PR ✅ 

Good job @daisybytes!

File: src/clj_sqs_extended/core.clj [hunk 17]
```diff
@@ -72,43 +73,58 @@
...
-     :or   {wait-time 0}
+     :or   {format :transit
+            wait-time 0}
      :as   opts}]
    (letfn [(extract-relevant-keys [message]
              (-> (bean message)
                  (select-keys [:messageId :receiptHandle :body])))]
      (let [request (doto (ReceiveMessageRequest. url)
                      (.setWaitTimeSeconds (int wait-time))
                      (.setMaxNumberOfMessages (int 1)))
```
Comment (Quantisan) [inline src/clj_sqs_extended/core.clj:121]: 1 is int already. you can drop the `int` @daisybytes. Also leave a design choice note as to why you're hardcoding this value to 1.

File: src/clj_sqs_extended/core.clj (hunk 6 above)
Comment (ghost) [reply src/clj_sqs_extended/core.clj]: Having this available on receive-message only might not interfere too much with devops and infrastructure configuration since it doesn't change the entire queue? Even though I have basically no real-life experience with sending SQS messages between real-world applications I thought this might come in handy.

File: src/clj_sqs_extended/core.clj (hunk 16 above)
Comment (ghost) [reply src/clj_sqs_extended/core.clj]: Stupid miss.

File: src/clj_sqs_extended/core.clj (hunk 6 above)
Comment (Quantisan) [reply src/clj_sqs_extended/core.clj]: that's true. this would only change these particular messages. but from my experience, I find having this option available at two different places confusing as a user.

File: test/clj_sqs_extended/core_test.clj (hunk 5 above)
Comment (ghost) [reply test/clj_sqs_extended/core_test.clj]: Extended to 5.

File: src/clj_sqs_extended/core.clj (hunk 17 above)
Comment (ghost) [reply src/clj_sqs_extended/core.clj:121]: Hmm.
(type 1)
=> java.lang.Long

File: test/clj_sqs_extended/core_test.clj (hunk 7 above)
Comment (ghost) [reply test/clj_sqs_extended/core_test.clj]: I'm really sorry, but I don't get what you'd like to minimize exactly.

Comment (Quantisan) [reply test/clj_sqs_extended/core_test.clj]: on second thought, nevermind @daisybytes

File: src/clj_sqs_extended/core.clj (hunk 17 above)
Comment (Quantisan) [reply src/clj_sqs_extended/core.clj:121]: true. does it cause a problem to the client library passing a Long 1? Usually it's ok. Clojure/Java interop should be forgiving on types.

Comment (ghost) [reply src/clj_sqs_extended/core.clj:121]: Unfortunately the underlying library is very picky here and throws an exception without the additional ugly int.

Comment (Quantisan) [reply src/clj_sqs_extended/core.clj:121]: @daisybytes ahh ok

File: src/clj_sqs_extended/core.clj (hunk 6 above)
Comment (ghost) [reply src/clj_sqs_extended/core.clj]: "AAAND It's Gone" :wink:

File: src/clj_sqs_extended/core.clj [hunk 18]
```diff
@@ -65,57 +67,64 @@
...
+            wait-time 0}}]
    (letfn [(extract-relevant-keys [message]
              (-> (bean message)
                  (select-keys [:messageId :receiptHandle :body])))]
      (let [request (doto (ReceiveMessageRequest. url)
                      (.setWaitTimeSeconds (int wait-time))
+                     ;;
+                     ;; WATCHOUT: For this library's use-case in combination with core.async
+                     ;;           channels later the number of messages returned gets hardcoded
+                     ;;           to 1 here.
```
Comment (Quantisan) [inline src/clj_sqs_extended/core.clj]: this warning isn't clear for reader without context. I'd rephrase it as,

`this is a design choice to read one message at a time from the queue`

File: src/clj_sqs_extended/core.clj [hunk 19]
```diff
@@ -65,57 +67,64 @@
...
      :or   {format :transit
-            wait-time 0}
-     :as   opts}]
+            wait-time 0}}]
    (letfn [(extract-relevant-keys [message]
              (-> (bean message)
                  (select-keys [:messageId :receiptHandle :body])))]
      (let [request (doto (ReceiveMessageRequest. url)
                      (.setWaitTimeSeconds (int wait-time))
+                     ;;
```
Comment (Quantisan) [inline src/clj_sqs_extended/core.clj]: we don't need extra white space as these are implementation warnings so in most likelihood, only whoever is patching this particular line would need to worry about it. thus, they'll be reading this in detail anyway.

File: src/clj_sqs_extended/serdes.clj [hunk 20]
```diff
@@ -39,13 +39,17 @@
     (= format :transit) (transit-write out)
     (= format :json) (json/generate-string out)
     (= format :raw) out
-    :else out))
+    :else (throw (ex-info "Format not supported"
```
Comment (Quantisan) [inline src/clj_sqs_extended/serdes.clj]: error message should guide user on how to fix the error. i.e. `(ex-info (format "Only %s formats are supported. We received %s." supported-formats format))`

File: src/clj_sqs_extended/serdes.clj [hunk 21]
```diff
@@ -39,13 +39,17 @@
...
 (defn deserialize
   [in format]
   (cond
     (= format :transit) (transit-read in)
     (= format :json) (json/parse-string in true)
     (= format :raw) in
-    :else in))
+    :else (throw (ex-info "Format not supported"
+                          {:supported [:transit :json]
+                           :requested format}))))
```
Comment (Quantisan) [inline src/clj_sqs_extended/serdes.clj]: since this error is same as above, I'd refactor this out into a private fn

File: test/clj_sqs_extended/core_test.clj [hunk 22]
```diff
@@ -8,52 +8,67 @@
...
+  [{:quote "State. You're doing it wrong."
+    :by    "Rich Hickey"}
+   {:quote "Nothing says 'Screw You!' like a DSL."
+    :by    "Stuart Halloway"}
+   {:quote "I don't care if it works on your machine! We are not shipping your machine!"
+    :by    "Vidiu Platon"}
+   {:quote "If you can't beat your computer at chess, try kickboxing."
+    :by    "Claudio E. Montero"}
+   {:quote "I would rather be without a state than without a voice."
+    :by    "Edward Snowden"}])
```
Comment (Quantisan) [inline test/clj_sqs_extended/core_test.clj]: it's fun with one or two of these. but this is now taking up 10 lines of code for no reason. can we just generate some random messages instead?

File: test/clj_sqs_extended/core_test.clj [hunk 23]
```diff
@@ -8,52 +8,67 @@
...
         (let [message (sqs-ext/receive-message client url {:format format})]
-          (is (= test-message-1 (:body message))))))))
+          (is (= test-message (:body message))))))))
 
 (deftest can-auto-delete-message
   (testing "Auto-Deleting a single message after receiving it"
     (let [client (fixtures/localstack-sqs)
-          url (fixtures/test-queue-url)]
+          url (fixtures/test-queue-url)
+          test-message (rand-nth test-messages)]
```
Comment (Quantisan) [inline test/clj_sqs_extended/core_test.clj]: good re-use of `test-messages`. I would statically pick one so that there's no randomness in this test that doesn't need it. For repeatability's sake.

File: test/clj_sqs_extended/core_test.clj [hunk 24]
```diff
@@ -8,52 +8,67 @@
...
+                                     url
+                                     message
+                                     (helpers/random-string-with-length 5)
+                                     {:format format}))
         (let [message (sqs-ext/receive-message client url {:format format})]
-          (is (= test-message-1 (:body message))))
+          (is (= (nth test-messages 0) (:body message))))
         (let [message (sqs-ext/receive-message client url {:format format})]
-          (is (= test-message-2 (:body message))))))))
+          (is (= (nth test-messages 1) (:body message))))
```
Comment (Quantisan) [inline test/clj_sqs_extended/core_test.clj]: this is repetitive. keep track of an `i` index in your doseq and so similar to above as L56 to iterate the reads through.

File: src/clj_sqs_extended/core.clj [hunk 25]
```diff
@@ -129,11 +128,6 @@
            response (.receiveMessage sqs-client request)
            message (->> (.getMessages response) (first) (extract-relevant-keys))
            payload (serdes/deserialize (:body message) format)]
-       (when visibility-timeout
```
Comment (Quantisan) [inline src/clj_sqs_extended/core.clj:107]: thanks

File: test/clj_sqs_extended/core_test.clj [hunk 26]
```diff
@@ -71,20 +71,3 @@
             desired-length (count test-message-larger-than-256kb)
             response-length (->> (:body message) (count))]
         (is (= desired-length response-length))))))
-
-(deftest can-receive-again-after-visibility-timeout-expired
```
Comment (Quantisan) [inline test/clj_sqs_extended/core_test.clj:57]: 👍

File: src/clj_sqs_extended/serdes.clj [hunk 27]
```diff
@@ -63,22 +62,23 @@
         reader (transit/reader in :json read-handlers)]
     (transit/read reader)))
 
+(defn- unsupported-format-exception
+  [got]
+  (ex-info (format "Only %s formats are supported. We received %s."
+                   [:transit :json] got) {}))
```
Comment (Quantisan) [inline src/clj_sqs_extended/serdes.clj]: new line with `{}` since it's the 2nd arg

also, perhaps pass `{:cause :unsupported-serde-format}` instead of empty map

File: test/clj_sqs_extended/core_test.clj [hunk 28]
```diff
@@ -33,11 +24,11 @@
         (let [message (sqs-ext/receive-message client url {:format format})]
           (is (= test-message (:body message))))))))
 
-(deftest can-auto-delete-message
+(deftest can-auto-delete-messageX
```
Comment (Quantisan) [inline test/clj_sqs_extended/core_test.clj]: what's `messageX`?

File: src/clj_sqs_extended/serdes.clj (hunk 27 above)
Comment (Quantisan) [reply src/clj_sqs_extended/serdes.clj]: @daisybytes you can merge after this change

File: test/clj_sqs_extended/core_test.clj (hunk 28 above)
Comment (ghost) [reply test/clj_sqs_extended/core_test.clj]: Damn. That happens when you slip while using your run-tests shortcut :sob:
